cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/PreviewServerModule.java

```diff
@@ -0,0 +1,113 @@
+import co.cask.cdap.config.guice.ConfigStoreModule;
+import co.cask.cdap.gateway.handlers.CommonHandlers;
+import co.cask.cdap.gateway.handlers.PreviewHttpHandler;
+import co.cask.cdap.internal.app.deploy.pipeline.AppDeploymentInfo;
+import co.cask.cdap.internal.app.deploy.pipeline.ApplicationWithPrograms;
+import co.cask.cdap.internal.app.preview.DefaultPreviewManager;
+import co.cask.cdap.internal.app.runtime.schedule.NoopScheduler;
+import co.cask.cdap.internal.app.runtime.schedule.Scheduler;
+import co.cask.cdap.internal.app.services.ProgramLifecycleService;
+import co.cask.cdap.internal.app.store.DefaultStore;
+import co.cask.cdap.internal.pipeline.SynchronousPipelineFactory;
+import co.cask.cdap.pipeline.PipelineFactory;
+import co.cask.http.HttpHandler;
+import com.google.inject.Exposed;
+import com.google.inject.Module;
+import com.google.inject.PrivateModule;
+import com.google.inject.Provides;
+import com.google.inject.Scopes;
+import com.google.inject.TypeLiteral;
+import com.google.inject.assistedinject.FactoryModuleBuilder;
+import com.google.inject.multibindings.Multibinder;
+import com.google.inject.name.Named;
+import com.google.inject.name.Names;
+import com.google.inject.util.Modules;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+
+/**
+ * Preview server modules.
+ */
+public class PreviewServerModule {
+
+  public Module getPreviewModules() {
+    return Modules.combine(new PreviewServiceModule(),
+                           new ConfigStoreModule().getStandaloneModule());
+  }
+
+  private static final class PreviewServiceModule extends PrivateModule {
+    @Override
+    protected void configure() {
+      bind(PipelineFactory.class).to(SynchronousPipelineFactory.class);
+      expose(PipelineFactory.class);
+
+      install(
+        new FactoryModuleBuilder()
+          .implement(new TypeLiteral<Manager<AppDeploymentInfo, ApplicationWithPrograms>>() {
+                     },
+                     new TypeLiteral<PreviewApplicationManager<AppDeploymentInfo, ApplicationWithPrograms>>() {
+                     })
+          .build(new TypeLiteral<ManagerFactory<AppDeploymentInfo, ApplicationWithPrograms>>() {
+          })
+      );
+
+      bind(Store.class).to(DefaultStore.class);
+      expose(Store.class);
+
+      bind(ProgramLifecycleService.class).in(Scopes.SINGLETON);
+
+      bind(PreviewServer.class).in(Scopes.SINGLETON);
+      expose(PreviewServer.class);
+
+      bind(PreviewManager.class).to(DefaultPreviewManager.class).in(Scopes.SINGLETON);
+      expose(PreviewManager.class);
+
+      Multibinder<HttpHandler> handlerBinder = Multibinder.newSetBinder(
+        binder(), HttpHandler.class, Names.named(Constants.Preview.HANDLERS_BINDING));
+
+      CommonHandlers.add(handlerBinder);
+      handlerBinder.addBinding().to(PreviewHttpHandler.class);
+
+
+      bind(Scheduler.class).to(NoopScheduler.class);
+      expose(Scheduler.class);
+    }
+
+    @Provides
+    @Exposed
```

Comment:
co.cask.cdap.app.guice.InMemoryProgramRunnerModule.DiscoveryServiceAnnouncer announces on "app.bind.address", so we need this to be exposed. let me know if it can be done by another way.